Also get the content headers when making StaticAssets calls

pbiggar opened the pull request:
This fixes the StaticAssets issue.

I want to add tests before merging though.

pbiggar commented:
The fix to static assets was to pass through the content-headers as well as the other headers. We had code to do this for LibHttpClient but forgot it here, so I refactored to make it harder to forget.

I added tests and refactored the test suite to make it easier to do so. Because static assets test implicitly have the canvasName in them (that is where the first hash is generated from), it was necessary to stop using random canvas names for these particular tests, so I added `ExactCanvasName`. I also made it so we could have multiple annotations following a `with` on a test or test group, which was needed for this test. Finally, I added a way to add a static deploy to the DB.

I uploaded the 3 files in static-assets-tests to the cloud storage bucket pointed to by the `test-static-assets-deploy` canvas. This is at `https://test-static-assets-deploy.darksa.com/u0n38z-mv-t0tkvnwiqyvankw6w/abcdef1234/index.html`. I noticed that index.html would 404 sometimes, hopefully that's just a brief hiccup and doesn't stick around.

As well as adding tests for the remaining functions (listed [here](https://github.com/darklang/dark/issues/3658#issuecomment-1091041907)), the tests currently have a major race condition. Since each test uses the same canvas, when each test is cleared between test runs, they clean the DB. If another test is running using the same canvas, the tests will fail. I thought about creating a canvas for each test, but it would also need a bucket for each test, which would be frustrating. but maybe that's the only way.

To get this over the line we need to:
- make tests work consistently (eliminate the race condition)
- add tests for the other functions

StachuDotNet commented on scripts/formatting/format:
```diff
@@ -125,6 +125,8 @@ all_prettier_files() {
     -o -path "./fsharp-backend/.paket" \
     -o -path "fsharp-backend/Build" \
     -o -path "./fsharp-backend/Build" \
+    -o -path "fsharp-backend/tests/staticassets-tests" \
+    -o -path "./fsharp-backend/tests/staticassets-tests" \
```
is this to deal with OS inconsistencies?

pbiggar replied:
You mean the fact that it's there twice? I don't remember. I'm cargo culting it a bit here.

StachuDotNet replied:
yeah that's what I meant; OK, np

StachuDotNet commented on fsharp-backend/tests/Tests/LibExecution.Tests.fs:
```diff
@@ -51,15 +54,40 @@ let setUpWorkers meta workers =
 
+let setupStaticAssets
+  (meta : Canvas.Meta)
+  (deployHashes : List<string>)
```
nitpick that we can move elsewhere if better:

built-in types like `list` are generally (at least, by docs I've read, incl. built-in types like `list` are generally (at least, by docs I've read) used in the suffix notation (`x list` rather than `List<x>`)
I'm totally fine getting away from that, but we should mark that in the coding guide.) used in the suffix notation (`x list` rather than `List<x>`), standardize to that, and (as we see it) adjust to use that generally.

There's a LinkedList/Generic.List type that also exists in .NET

upon seeing List<x>, I think many devs brains would think of that mutable array list thing first, esp. if any C# experience

pbiggar commented:
Am continuing with this. I just remembered that we can use `testSequencedGroup` to avoid races.

pbiggar commented:
Summary of whole PR:

The fix to static assets was to pass through the content-headers as well as the other headers. We had code to do this for LibHttpClient but forgot it here, so I refactored to make it harder to forget.

I added tests and refactored the test suite to make it easier to do so. Because static assets test implicitly have the canvasName in them (that is where the first hash is generated from), it was necessary to stop using random canvas names for these particular tests, so I added ExactCanvasName. I also made it so we could have multiple annotations following a with on a test or test group, which was needed for this test. Finally, I added a way to add a static deploy to the DB.

I uploaded the 3 files in static-assets-tests to the cloud storage bucket pointed to by the test-static-assets-deploy canvas. This is at https://test-static-assets-deploy.darksa.com/u0n38z-mv-t0tkvnwiqyvankw6w/abcdef1234/index.html. (Note that this would occasionally fail, not consistently, probably an eventual consistency thing. Anyway, it has stopped).

Since each test uses the same canvas, I had to run all the tests in serial. This slows down the test suite so we may need to split this file up later to speed it up, but fine for now.


Also:

- created a specific config var (STATIC_ASSETS_SALT_SUFFIX) instead of using ENV_DISPLAY_NAME (this was the make tests work in CI, but it's good practice anyway)

- add `Test.regexReplace` (I ended up not using it in the end, but nice to have)

- Add `Test.httpResponseHeaders`, `Test.httpResponseStatusCode`, and  `Test.httpResponseBody`. There isn't another way to get values out of a httpresponse

- Add `Test.intArrayToBytes` - allow creating a DBytes value in a test

- Fix error message when not a 200 code (we would throw the exception then catch it immediately)

- better error message when there's no deploy hash

- Change which headers are removed in the F# version:
  - remove Content-length cause asp.net likes to set it
  - remove Server as nginx overrode that before, and we don't want to set the Google server header as our server header (it's `UploadServer`)
  - allow Cache-Control (we tried to disable it before but it didn't work, so this keeps things the same)

StachuDotNet commented on backend/liblegacyfuzzing/fuzzing.ml:
```diff
@@ -472,6 +472,83 @@ let fns : Types.RuntimeT.fn list =
+            ( match response with
+            | Redirect _, _ ->
+                Dval.dint 302
+            | Response (code, headers), _ ->
```
```suggestion
            | Response (code, _), _ ->
```
nitpick

StachuDotNet commented on fsharp-backend/tests/Tests/LibExecution.Tests.fs:
```diff
@@ -441,38 +504,48 @@ let fileTests () : Test =
-    testList $"Tests from {filename}" allTests)
+    let tests = testList $"Tests from {filename}" fileTests
+    // Staticassets.tests use a combination of ExactCanvasName and
+    // StaticAssetsDeployHash, which can be racy. See comment in staticassets.tests.
```
replace 'racy' with less ambiguous term like "Race conditions can cause problems with these tests" here and in `staticassets.tests`?

Wasn't sure if 'racy' meant 'involves race conditions' or 'annoying/dicy' or something else, at first read

StachuDotNet commented on fsharp-backend/tests/httptestfiles/serve-latest.test:
```diff
@@ -0,0 +1,50 @@
+[http-handler GET /]
+StaticAssets.serve_v1_ster "abcdef1234" "index.html"
+
+[canvas-name test-static-assets-deploy]
```
Not a big deal, but I'd expect this at the top of the file for single-file tests

StachuDotNet commented on fsharp-backend/tests/testfiles/staticassets.tests:
```diff
@@ -0,0 +1,547 @@
+
+// baseUrlFor
+[test.baseUrlFor] with ExactCanvasName test-baseurlfor-0
+StaticAssets.baseUrlFor_v0 "abcdefghi" = "https://test-baseurlfor-0.darksa.com/gd57quywk2db6zncs3h2hxgfdtc/abcdefghi"
```
where does `gd57quywk2db6zncs3h2hxgfdtc` come from?

pbiggar replied:
These are all hashes made from the canvas name

https://github.com/darklang/dark/blob/d88146e9d404ce8aa59e82096f15b044f130a8cf/fsharp-backend/src/LibBackend/StaticAssets.fs#L67

StachuDotNet commented:
Tests to consider adding to `staticassets.tests`:
- (easy?) for `serve` and similar fns, tests where we _have_ a deploy has, but reference a different one
- (more complicated) for `serveLatest`, deploy _twice_, and ensure the latest is the one served (unless I missed this)?

StachuDotNet commented:
Done another round of review @pbiggar

pbiggar commented on fsharp-backend/tests/testfiles/staticassets.tests:
```diff
@@ -0,0 +1,547 @@
+// Serve
+// ----------------------------
+
+[fn.ocaml200HtmlHeaders i:int]
```
Yeah, exactly. As far as I recall, we can't call functions without arguments in these tests.

I'll rename the parameters to `_`

pbiggar commented:
> Tests to consider adding to `staticassets.tests`:
> 
> * (easy?) for `serve` and similar fns, tests where we _have_ a deploy has, but reference a different one


> * (more complicated) for `serveLatest`, deploy _twice_, and ensure the latest is the one served (unless I missed this)?

Yes, I can add two deploy hashes easily enough.
